htsget-test/src/http/concat.rs

```diff
@@ -0,0 +1,210 @@
+use base64::engine::general_purpose;
+use base64::Engine;
+use futures::future::join_all;
+use futures::{Stream, TryStreamExt};
+use htsget_config::types::{Class, Format, Response, Url};
+use http::{HeaderMap, HeaderName, HeaderValue};
+use noodles::{bam, bcf, bgzf, cram, fasta, vcf};
+use reqwest::Client;
+use std::future::Future;
+use std::io;
+use std::path::Path;
+use std::str::FromStr;
+use tokio::fs::File;
+use tokio::io::AsyncReadExt;
+
+/// A response concatenator which concatenates url tickets.
+#[derive(Debug)]
+pub struct ConcatResponse {
+  response: Response,
+  class: Class,
+}
+
+impl ConcatResponse {
+  /// Create a new response concatenator.
+  pub fn new(response: Response, class: Class) -> Self {
+    Self { response, class }
+  }
+
+  /// Get the inner response.
+  pub fn into_inner(self) -> Response {
+    self.response
+  }
+
+  /// Get the inner response.
+  pub fn response(&self) -> &Response {
+    &self.response
+  }
+
+  /// Concatenate a response into the bytes represented by the url ticket with a file path
+  pub async fn concat_from_file_path(self, path: impl AsRef<Path>) -> ReadRecords {
+    let file = File::open(path).await.unwrap();
+    self.concat_from_file(file).await
+  }
+
+  /// Concatenate a response into the bytes represented by the url ticket with file data.
+  pub async fn concat_from_file(self, mut file: File) -> ReadRecords {
+    let mut bytes = vec![];
+    file.read_to_end(&mut bytes).await.unwrap();
+
+    self.concat_from_bytes(bytes.as_slice()).await
+  }
+
+  /// Concatentate a response into bytes using a reqwest client.
+  pub async fn concat_from_client(self, client: &Client) -> ReadRecords {
+    let merged_bytes = join_all(self.response.urls.into_iter().map(|url| {
+      Self::url_to_bytes(url, |url| async move {
+        client
+          .get(url.url.as_str())
+          .headers(HeaderMap::from_iter(
+            url
+              .headers
+              .unwrap_or_default()
+              .into_inner()
+              .into_iter()
+              .map(|(key, value)| {
+                (
+                  HeaderName::from_str(&key).unwrap(),
+                  HeaderValue::from_str(&value).unwrap(),
+                )
+              }),
+          ))
+          .send()
+          .await
+          .unwrap()
```

Comment:
I'd return `Result` to avoid those `unwrap()s`?

Reply:
Yeah that's a good call. At some point it will be good to convert any `unwraps` to `Results` inside `htsget-test` too.